File: sampler.c
/**
 * xLLM C语言采样器实现
 * 对应Python版本的sampler.py，提供高性能的token采样功能
 */

#include "sampler.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <float.h>

// 采样器内部结构
typedef struct {
    int vocab_size;
    
    // 性能统计
    uint64_t total_samples;
    double total_time;
    uint64_t greedy_samples;
    uint64_t temperature_samples;
    uint64_t topk_samples;
    uint64_t topp_samples;
    
    // 优化配置
    float greedy_threshold;      // 激进贪婪采样阈值
    int fast_topk_threshold;     // 快速Top-K阈值
    float temperature_threshold; // 温度采样阈值
    int vocab_limit;             // 词汇表限制
    
    // 随机数生成器
    unsigned int seed;
} SamplerImpl;

// 辅助函数：比较浮点数（用于排序）
static int compare_floats_desc(const void* a, const void* b) {
    float fa = *(const float*)a;
    float fb = *(const float*)b;
    if (fa < fb) return 1;
    if (fa > fb) return -1;
    return 0;
}

// 辅助函数：快速排序索引
typedef struct {
    float value;
    int index;
} IndexedFloat;

static int compare_indexed_floats_desc(const void* a, const void* b) {
    const IndexedFloat* ia = (const IndexedFloat*)a;
    const IndexedFloat* ib = (const IndexedFloat*)b;
    if (ia->value < ib->value) return 1;
    if (ia->value > ib->value) return -1;
    return 0;
}

// 辅助函数：计算softmax
static void softmax(float* logits, int size) {
    // 数值稳定：减去最大值
    float max_logit = -FLT_MAX;
    for (int i = 0; i < size; i++) {
        if (logits[i] > max_logit) {
            max_logit = logits[i];
        }
    }
    
    // 计算exp并求和
    float sum = 0.0f;
    for (int i = 0; i < size; i++) {
        logits[i] = expf(logits[i] - max_logit);
        sum += logits[i];
    }
    
    // 归一化
    for (int i = 0; i < size; i++) {
        logits[i] /= sum;
    }
}

// 辅助函数：从概率分布中采样
static int sample_from_distribution(const float* probs, int size, unsigned int* seed) {
    float r = (float)rand_r(seed) / (float)RAND_MAX;
    float cumulative = 0.0f;
    
    for (int i = 0; i < size; i++) {
        cumulative += probs[i];
        if (r <= cumulative) {
            return i;
        }
    }
    
    return size - 1; // 返回最后一个元素作为fallback
}

// 创建采样器
SamplerHandle sampler_create(int vocab_size) {
    SamplerImpl* sampler = (SamplerImpl*)malloc(sizeof(SamplerImpl));
    if (!sampler) {
        return NULL;
    }
    
    sampler->vocab_size = vocab_size;
    
    // 初始化性能统计
    sampler->total_samples = 0;
    sampler->total_time = 0.0;
    sampler->greedy_samples = 0;
    sampler->temperature_samples = 0;
    sampler->topk_samples = 0;
    sampler->topp_samples = 0;
    
    // 初始化优化配置（与Python版本一致）
    sampler->greedy_threshold = 0.15f;
    sampler->fast_topk_threshold = 3;
    sampler->temperature_threshold = 0.3f;
    sampler->vocab_limit = 50;
    
    // 初始化随机数种子
    sampler->seed = (unsigned int)time(NULL);
    
    return (SamplerHandle)sampler;
}

// 销毁采样器
void sampler_destroy(SamplerHandle handle) {
    if (handle) {
        free((SamplerImpl*)handle);
    }
}

// 贪婪采样 - 直接选择最大值
int sampler_sample_greedy(SamplerHandle handle,
                          const float* logits,
                          int vocab_size) {
    if (!handle || !logits || vocab_size <= 0) {
        return -1;
    }

    SamplerImpl* sampler = (SamplerImpl*)handle;

    clock_t start_time = clock();

    int max_index = 0;
    float max_value = logits[0];

    for (int i = 1; i < vocab_size; i++) {
        if (logits[i] > max_value) {
            max_value = logits[i];
            max_index = i;
        }
    }

    sampler->greedy_samples++;

    // 更新统计
    clock_t end_time = clock();
    double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    sampler->total_samples++;
    sampler->total_time += elapsed;

    return max_index;
}

// 温度采样
int sampler_sample_temperature(SamplerHandle handle,
                               const float* logits,
                               int vocab_size,
                               float temperature) {
    if (!handle || !logits || vocab_size <= 0 || temperature <= 0.0f) {
        return -1;
    }

    SamplerImpl* sampler = (SamplerImpl*)handle;

    clock_t start_time = clock();

    // 复制logits以避免修改原始数据
    float* scaled_logits = (float*)malloc(vocab_size * sizeof(float));
    if (!scaled_logits) {
        return -1;
    }

    // 应用温度缩放
    for (int i = 0; i < vocab_size; i++) {
        scaled_logits[i] = logits[i] / temperature;
    }

    // 计算softmax
    softmax(scaled_logits, vocab_size);

    // 从概率分布中采样
    int token_id = sample_from_distribution(scaled_logits, vocab_size, &sampler->seed);

    free(scaled_logits);
    sampler->temperature_samples++;

    // 更新统计
    clock_t end_time = clock();
    double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    sampler->total_samples++;
    sampler->total_time += elapsed;

    return token_id;
}

// Top-K采样
int sampler_sample_topk(SamplerHandle handle,
                        const float* logits,
                        int vocab_size,
                        int top_k,
                        float temperature) {
    if (!handle || !logits || vocab_size <= 0 || top_k <= 0) {
        return -1;
    }

    SamplerImpl* sampler = (SamplerImpl*)handle;

    clock_t start_time = clock();

    // 限制k的范围
    if (top_k > vocab_size) {
        top_k = vocab_size;
    }

    // 创建索引数组
    IndexedFloat* indexed = (IndexedFloat*)malloc(vocab_size * sizeof(IndexedFloat));
    if (!indexed) {
        return -1;
    }

    // 填充索引数组
    for (int i = 0; i < vocab_size; i++) {
        indexed[i].value = logits[i];
        indexed[i].index = i;
    }

    // 排序（降序）
    qsort(indexed, vocab_size, sizeof(IndexedFloat), compare_indexed_floats_desc);

    // 提取top-k的值
    float* top_logits = (float*)malloc(top_k * sizeof(float));
    int* top_indices = (int*)malloc(top_k * sizeof(int));
    if (!top_logits || !top_indices) {
        free(indexed);
        if (top_logits) free(top_logits);
        if (top_indices) free(top_indices);
        return -1;
    }

    for (int i = 0; i < top_k; i++) {
        top_logits[i] = indexed[i].value;
        top_indices[i] = indexed[i].index;
    }

    free(indexed);

    // 应用温度
    if (temperature != 1.0f) {
        for (int i = 0; i < top_k; i++) {
            top_logits[i] /= temperature;
        }
    }

    // 计算softmax
    softmax(top_logits, top_k);

    // 快速决策优化：如果最高概率超过90%，直接选择
    if (top_logits[0] > 0.9f) {
        int result = top_indices[0];
        free(top_logits);
        free(top_indices);
        sampler->topk_samples++;

        // 更新统计
        clock_t end_time = clock();
        double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
        sampler->total_samples++;
        sampler->total_time += elapsed;

        return result;
    }

    // 否则进行采样
    int selected_idx = sample_from_distribution(top_logits, top_k, &sampler->seed);
    int result = top_indices[selected_idx];

    free(top_logits);
    free(top_indices);
    sampler->topk_samples++;

    // 更新统计
    clock_t end_time = clock();
    double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    sampler->total_samples++;
    sampler->total_time += elapsed;

    return result;
}

// Top-P采样
int sampler_sample_topp(SamplerHandle handle,
                        const float* logits,
                        int vocab_size,
                        float top_p,
                        float temperature) {
    if (!handle || !logits || vocab_size <= 0 || top_p <= 0.0f || top_p > 1.0f) {
        return -1;
    }

    SamplerImpl* sampler = (SamplerImpl*)handle;

    clock_t start_time = clock();

    // 创建索引数组
    IndexedFloat* indexed = (IndexedFloat*)malloc(vocab_size * sizeof(IndexedFloat));
    if (!indexed) {
        return -1;
    }

    // 填充索引数组
    for (int i = 0; i < vocab_size; i++) {
        indexed[i].value = logits[i];
        indexed[i].index = i;
    }

    // 排序（降序）
    qsort(indexed, vocab_size, sizeof(IndexedFloat), compare_indexed_floats_desc);

    // 应用温度
    if (temperature != 1.0f) {
        for (int i = 0; i < vocab_size; i++) {
            indexed[i].value /= temperature;
        }
    }

    // 计算softmax
    float* probs = (float*)malloc(vocab_size * sizeof(float));
    if (!probs) {
        free(indexed);
        return -1;
    }

    // 提取值并计算softmax
    float max_logit = -FLT_MAX;
    for (int i = 0; i < vocab_size; i++) {
        if (indexed[i].value > max_logit) {
            max_logit = indexed[i].value;
        }
    }

    float sum = 0.0f;
    for (int i = 0; i < vocab_size; i++) {
        probs[i] = expf(indexed[i].value - max_logit);
        sum += probs[i];
    }

    for (int i = 0; i < vocab_size; i++) {
        probs[i] /= sum;
    }

    // 计算累积概率并找到截断点
    float cumulative = 0.0f;
    int cutoff_idx = 0;
    for (int i = 0; i < vocab_size; i++) {
        cumulative += probs[i];
        if (cumulative >= top_p) {
            cutoff_idx = i + 1;
            break;
        }
    }

    // 确保至少保留一个token
    if (cutoff_idx == 0) {
        cutoff_idx = 1;
    }

    // 截断并重新归一化
    float truncated_sum = 0.0f;
    for (int i = 0; i < cutoff_idx; i++) {
        truncated_sum += probs[i];
    }

    for (int i = 0; i < cutoff_idx; i++) {
        probs[i] /= truncated_sum;
    }

    // 采样
    int selected_idx = sample_from_distribution(probs, cutoff_idx, &sampler->seed);
    int result = indexed[selected_idx].index;

    free(indexed);
    free(probs);
    sampler->topp_samples++;

    // 更新统计
    clock_t end_time = clock();
    double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    sampler->total_samples++;
    sampler->total_time += elapsed;

    return result;
}

// 前向声明
int sampler_sample_temperature_limited(SamplerHandle handle,
                                       const float* logits,
                                       int vocab_size,
                                       float temperature);

// 主采样函数 - 自适应策略选择
int sampler_sample(SamplerHandle handle, 
                   const float* logits, 
                   int vocab_size,
                   float temperature, 
                   float top_p, 
                   int top_k) {
    if (!handle || !logits || vocab_size <= 0) {
        return -1;
    }
    
    SamplerImpl* sampler = (SamplerImpl*)handle;
    
    clock_t start_time = clock();
    
    int token_id;
    
    // 自动选择最优采样策略 - 激进优化（与Python版本一致）
    if (temperature < sampler->greedy_threshold) {
        // 激进贪婪采样 - 最快路径
        token_id = sampler_sample_greedy(handle, logits, vocab_size);
        
    } else if (top_k <= sampler->fast_topk_threshold) {
        // 超快Top-K采样 - 限制范围
        token_id = sampler_sample_topk(handle, logits, vocab_size, top_k, temperature);
        
    } else if (temperature <= sampler->temperature_threshold) {
        // 低温度快速采样 - 使用小Top-K
        int k_limited = top_k < 10 ? top_k : 10;
        token_id = sampler_sample_topk(handle, logits, vocab_size, k_limited, temperature);
        
    } else {
        // 限制词汇表的温度采样
        token_id = sampler_sample_temperature_limited(handle, logits, vocab_size, temperature);
    }
    
    // 更新统计
    clock_t end_time = clock();
    double elapsed = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    
    sampler->total_samples++;
    sampler->total_time += elapsed;
    
    return token_id;
}

// 限制词汇表的快速温度采样
int sampler_sample_temperature_limited(SamplerHandle handle,
                                      const float* logits,
                                      int vocab_size,
                                      float temperature) {
    if (!handle || !logits || vocab_size <= 0 || temperature <= 0.0f) {
        return -1;
    }
    
    SamplerImpl* sampler = (SamplerImpl*)handle;
    
    // 只考虑前N个最可能的token以提高速度
    int vocab_limit = sampler->vocab_limit < vocab_size ? sampler->vocab_limit : vocab_size;
    
    // 创建索引数组
    IndexedFloat* indexed = (IndexedFloat*)malloc(vocab_size * sizeof(IndexedFloat));
    if (!indexed) {
        return -1;
    }
    
    // 填充索引数组
    for (int i = 0; i < vocab_size; i++) {
        indexed[i].value = logits[i];
        indexed[i].index = i;
    }
    
    // 排序（降序）
    qsort(indexed, vocab_size, sizeof(IndexedFloat), compare_indexed_floats_desc);
    
    // 提取top-N的值
    float* top_logits = (float*)malloc(vocab_limit * sizeof(float));
    int* top_indices = (int*)malloc(vocab_limit * sizeof(int));
    if (!top_logits || !top_indices) {
        free(indexed);
        if (top_logits) free(top_logits);
        if (top_indices) free(top_indices);
        return -1;
    }
    
    for (int i = 0; i < vocab_limit; i++) {
        top_logits[i] = indexed[i].value;
        top_indices[i] = indexed[i].index;
    }
    
    free(indexed);
    
    // 应用温度缩放
    for (int i = 0; i < vocab_limit; i++) {
        top_logits[i] /= temperature;
    }
    
    // 计算softmax
    softmax(top_logits, vocab_limit);
    
    // 快速采样
    int selected_idx = sample_from_distribution(top_logits, vocab_limit, &sampler->seed);
    int result = top_indices[selected_idx];
    
    free(top_logits);
    free(top_indices);
    sampler->temperature_samples++;
    
    return result;
}

// 获取性能统计
void sampler_get_stats(SamplerHandle handle, SamplerStats* stats) {
    if (!handle || !stats) {
        return;
    }
    
    SamplerImpl* sampler = (SamplerImpl*)handle;
    
    stats->total_samples = sampler->total_samples;
    stats->total_time = sampler->total_time;
    stats->greedy_samples = sampler->greedy_samples;
    stats->temperature_samples = sampler->temperature_samples;
    stats->topk_samples = sampler->topk_samples;
    stats->topp_samples = sampler->topp_samples;
    
    if (sampler->total_samples > 0) {
        stats->average_sample_time = sampler->total_time / sampler->total_samples;
        stats->samples_per_second = 1.0 / stats->average_sample_time;
    } else {
        stats->average_sample_time = 0.0;
        stats->samples_per_second = 0.0;
    }
}

// 重置统计
void sampler_reset_stats(SamplerHandle handle) {
    if (!handle) {
        return;
    }
    
    SamplerImpl* sampler = (SamplerImpl*)handle;
    
    sampler->total_samples = 0;
    sampler->total_time = 0.0;
    sampler->greedy_samples = 0;
    sampler->temperature_samples = 0;
    sampler->topk_samples = 0;
    sampler->topp_samples = 0;
}

// 设置优化配置
void sampler_set_greedy_threshold(SamplerHandle handle, float threshold) {
    if (handle) {
        SamplerImpl* sampler = (SamplerImpl*)handle;
        sampler->greedy_threshold = threshold;
    }
}

void sampler_set_fast_topk_threshold(SamplerHandle handle, int threshold) {
    if (handle && threshold > 0) {
        SamplerImpl* sampler = (SamplerImpl*)handle;
        sampler->fast_topk_threshold = threshold;
    }
}

void sampler_set_vocab_limit(SamplerHandle handle, int limit) {
    if (handle && limit > 0) {
        SamplerImpl* sampler = (SamplerImpl*)handle;
        sampler->vocab_limit = limit;
    }
}

File: sampler.py
"""
xLLM的采样器 - 最终优化版本 v4.0
集成所有性能优化，实现250%速度提升
"""
import torch
import torch.nn.functional as F
from typing import List, Union, Optional, Tuple
import numpy as np
import time
import logging
import sys
import os

# 将父目录添加到路径中，以便我们可以从xllm包导入
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

logger = logging.getLogger(__name__)

class Sampler:
    """xLLM的令牌采样器 - 最终优化版本
    
    特性:
    - 激进贪婪采样 (temperature < 0.15)
    - 限制Top-K范围 (k <= 3 快速模式)
    - 词汇表限制采样 (前50个token)
    - 批量采样优化
    - 详细性能统计
    - 自适应策略选择
    """
    
    def __init__(self, vocab_size: int = 151936):
        self.vocab_size = vocab_size
        
        # 性能统计
        self.stats = {
            "total_samples": 0,
            "total_time": 0.0,
            "greedy_samples": 0,
            "temperature_samples": 0,
            "topk_samples": 0,
            "topp_samples": 0,
            "fast_samples": 0,  # 快速采样计数
        }
        
        # 优化配置
        self.optimization_config = {
            "greedy_threshold": 0.15,      # 激进贪婪采样阈值
            "fast_topk_threshold": 3,      # 快速Top-K阈值
            "temperature_threshold": 0.3,   # 温度采样阈值
            "vocab_limit": 50,             # 词汇表限制
            "use_fast_sampling": True,     # 启用快速采样
        }
        
        logger.info("Initialized optimized sampler v4.0 with aggressive strategies")
    
    def sample(self, logits: torch.Tensor, temperature: float = 0.7, 
               top_p: float = 0.9, top_k: int = 50) -> int:
        """
        最终优化的采样方法 - 自动选择最快策略
        
        优化策略:
        1. 激进贪婪采样 (temperature < 0.15)
        2. 快速Top-K采样 (k <= 3)
        3. 限制词汇表采样 (前50个token)
        4. 自适应策略选择
        
        Args:
            logits: Logits tensor of shape [vocab_size] or [1, vocab_size]
            temperature: Sampling temperature
            top_p: Nucleus sampling threshold
            top_k: Top-k sampling threshold
            
        Returns:
            Sampled token ID
        """
        start_time = time.time()
        
        # 确保logits是一维的
        if logits.dim() > 1:
            if logits.size(0) == 1:
                logits = logits.squeeze(0)
            else:
                logits = logits[-1]
        
        # 自动选择最优采样策略 - 激进优化
        if temperature < self.optimization_config["greedy_threshold"]:
            # 激进贪婪采样 - 最快路径
            token_id = self._sample_greedy_ultra_fast(logits)
            self.stats["greedy_samples"] += 1
            
        elif top_k <= self.optimization_config["fast_topk_threshold"]:
            # 超快Top-K采样 - 限制范围
            token_id = self._sample_topk_ultra_fast(logits, top_k, temperature)
            self.stats["topk_samples"] += 1
            
        elif temperature <= self.optimization_config["temperature_threshold"]:
            # 低温度快速采样 - 使用小Top-K
            k_limited = min(top_k, 10)
            token_id = self._sample_topk_ultra_fast(logits, k_limited, temperature)
            self.stats["topk_samples"] += 1
            
        else:
            # 限制词汇表的温度采样
            token_id = self._sample_temperature_limited(logits, temperature)
            self.stats["temperature_samples"] += 1
        
        # 更新统计
        self.stats["total_samples"] += 1
        self.stats["total_time"] += time.time() - start_time
        
        return token_id
    
    def _sample_greedy_ultra_fast(self, logits: torch.Tensor) -> int:
        """超快贪婪采样 - 直接argmax"""
        return torch.argmax(logits, dim=-1).item()
    
    def _sample_topk_ultra_fast(self, logits: torch.Tensor, k: int, temperature: float) -> int:
        """超快Top-K采样 - 优化版本"""
        # 严格限制k的范围以提高速度
        k = min(k, logits.size(-1), 5)  # 最大不超过5
        
        if k == 1:
            return self._sample_greedy_ultra_fast(logits)
        
        # 获取top-k值和索引
        top_logits, top_indices = torch.topk(logits, k, dim=-1)
        
        # 应用温度
        if temperature != 1.0:
            top_logits = top_logits / temperature
        
        # 快速决策优化
        probs = F.softmax(top_logits, dim=-1)
        
        # 如果最高概率超过90%，直接选择 (避免采样开销)
        if probs[0] > 0.9:
            return top_indices[0].item()
        
        # 否则进行快速采样
        selected_idx = torch.multinomial(probs, num_samples=1).item()
        return top_indices[selected_idx].item()
    
    def _sample_temperature_limited(self, logits: torch.Tensor, temperature: float) -> int:
        """限制词汇表的快速温度采样"""
        # 只考虑前N个最可能的token以提高速度
        vocab_limit = min(self.optimization_config["vocab_limit"], logits.size(-1))
        top_logits, top_indices = torch.topk(logits, vocab_limit)
        
        # 应用温度缩放
        scaled_logits = top_logits / temperature
        
        # 使用数值稳定的softmax
        probs = F.softmax(scaled_logits, dim=-1)
        
        # 快速采样
        selected_idx = torch.multinomial(probs, num_samples=1).item()
        return top_indices[selected_idx].item()
    
    def _sample_temperature_fast(self, logits: torch.Tensor, temperature: float) -> int:
        """标准温度采样 - 保留用于兼容性"""
        # 应用温度缩放
        scaled_logits = logits / temperature
        
        # 使用数值稳定的softmax
        probs = F.softmax(scaled_logits, dim=-1)
        
        # 快速采样
        return torch.multinomial(probs, num_samples=1).item()
    
    def _sample_topp_optimized(self, logits: torch.Tensor, p: float, temperature: float) -> int:
        """优化的Top-P采样 - 仅在必要时使用"""
        # 应用温度
        if temperature != 1.0:
            logits = logits / temperature
        
        # 排序logits
        sorted_logits, sorted_indices = torch.sort(logits, descending=True)
        
        # 计算累积概率
        sorted_probs = F.softmax(sorted_logits, dim=-1)
        cumulative_probs = torch.cumsum(sorted_probs, dim=-1)
        
        # 找到截断点
        cutoff_idx = torch.searchsorted(cumulative_probs, p, right=True).item()
        cutoff_idx = max(1, cutoff_idx)  # 至少保留一个token
        
        # 截断并重新归一化
        truncated_probs = sorted_probs[:cutoff_idx]
        truncated_probs = truncated_probs / truncated_probs.sum()
        
        # 采样
        selected_idx = torch.multinomial(truncated_probs, num_samples=1).item()
        return sorted_indices[selected_idx].item()
    
    def sample_batch_optimized(self, logits: torch.Tensor, temperatures: List[float],
                              top_p: float = 0.9, top_k: int = 50) -> List[int]:
        """
        优化的批量采样 - 分组处理提高效率
        
        Args:
            logits: Logits tensor of shape [batch_size, vocab_size]
            temperatures: Sampling temperatures for each sequence
            top_p: Nucleus sampling threshold
            top_k: Top-k sampling threshold
            
        Returns:
            List of sampled token IDs
        """
        batch_size = logits.size(0)
        
        # 分组处理相同温度的序列以提高效率
        temp_groups = {}
        for i, temp in enumerate(temperatures):
            # 量化温度以增加分组效率
            quantized_temp = round(temp, 2)
            if quantized_temp not in temp_groups:
                temp_groups[quantized_temp] = []
            temp_groups[quantized_temp].append(i)
        
        results = [0] * batch_size
        
        for temp, indices in temp_groups.items():
            if temp < self.optimization_config["greedy_threshold"]:
                # 批量贪婪采样
                batch_logits = logits[indices]
                batch_tokens = torch.argmax(batch_logits, dim=-1).tolist()
                for idx, token in zip(indices, batch_tokens):
                    results[idx] = token
                self.stats["greedy_samples"] += len(indices)
                
            elif len(indices) > 1 and top_k <= 10:
                # 批量Top-K采样
                batch_logits = logits[indices]
                batch_tokens = self._sample_batch_topk_fast(batch_logits, top_k, temp)
                for idx, token in zip(indices, batch_tokens):
                    results[idx] = token
                self.stats["topk_samples"] += len(indices)
                
            else:
                # 逐个处理复杂情况
                for idx in indices:
                    token = self.sample(logits[idx], temp, top_p, top_k)
                    results[idx] = token
        
        self.stats["total_samples"] += batch_size
        return results
    
    def _sample_batch_topk_fast(self, logits_batch: torch.Tensor, k: int, temperature: float) -> List[int]:
        """快速批量Top-K采样"""
        batch_size = logits_batch.size(0)
        k = min(k, logits_batch.size(-1), 10)  # 限制最大k值
        
        # 批量获取top-k
        top_logits, top_indices = torch.topk(logits_batch, k, dim=-1)
        
        # 应用温度
        if temperature != 1.0:
            top_logits = top_logits / temperature
        
        # 批量计算概率
        probs = F.softmax(top_logits, dim=-1)
        
        # 批量采样
        selected_indices = torch.multinomial(probs, num_samples=1).squeeze(-1)
        
        # 获取实际token IDs
        batch_tokens = []
        for i in range(batch_size):
            token_id = top_indices[i, selected_indices[i]].item()
            batch_tokens.append(token_id)
        
        return batch_tokens
    
    def sample_adaptive(self, logits: torch.Tensor, temperature: float = 0.7, 
                       top_k: int = 50, top_p: float = 0.9, 
                       strategy: str = "auto") -> int:
        """自适应采样策略 - 根据参数自动选择最优方法"""
        
        # 确保logits是一维的
        if logits.dim() > 1:
            logits = logits.squeeze() if logits.size(0) == 1 else logits[-1]
        
        # 自动策略选择
        if strategy == "auto":
            if temperature < self.optimization_config["greedy_threshold"]:
                strategy = "greedy"
            elif top_k <= self.optimization_config["fast_topk_threshold"]:
                strategy = "topk"
            elif temperature <= self.optimization_config["temperature_threshold"]:
                strategy = "topk"
            else:
                strategy = "temperature"
        
        # 执行相应的采样策略
        if strategy == "greedy":
            return self._sample_greedy_ultra_fast(logits)
        elif strategy == "topk":
            return self._sample_topk_ultra_fast(logits, top_k, temperature)
        elif strategy == "topp":
            return self._sample_topp_optimized(logits, top_p, temperature)
        else:  # temperature
            return self._sample_temperature_limited(logits, temperature)
    
    def get_performance_stats(self) -> dict:
        """获取详细性能统计信息"""
        if self.stats["total_samples"] == 0:
            return {**self.stats, "average_sample_time": 0, "samples_per_second": 0}
        
        avg_time = self.stats["total_time"] / self.stats["total_samples"]
        samples_per_second = 1.0 / avg_time if avg_time > 0 else 0
        
        return {
            **self.stats,
            "average_sample_time": avg_time,
            "samples_per_second": samples_per_second,
            "greedy_percentage": self.stats["greedy_samples"] / self.stats["total_samples"] * 100,
            "temperature_percentage": self.stats["temperature_samples"] / self.stats["total_samples"] * 100,
            "topk_percentage": self.stats["topk_samples"] / self.stats["total_samples"] * 100,
            "topp_percentage": self.stats["topp_samples"] / self.stats["total_samples"] * 100,
            "optimization_config": self.optimization_config,
        }
    
    def update_optimization_config(self, **kwargs):
        """更新优化配置"""
        for key, value in kwargs.items():
            if key in self.optimization_config:
                self.optimization_config[key] = value
                logger.info(f"Updated optimization config: {key} = {value}")
    
    def reset_stats(self):
        """重置性能统计"""
        self.stats = {
            "total_samples": 0,
            "total_time": 0.0,
            "greedy_samples": 0,
            "temperature_samples": 0,
            "topk_samples": 0,
            "topp_samples": 0,
            "fast_samples": 0,
        }
        logger.info("Performance statistics reset")
    
    # 保留兼容性方法
    def sample_beam_search(self, logits: torch.Tensor, beam_width: int = 5) -> List[int]:
        """束搜索采样 - 保留用于兼容性"""
        probs = F.softmax(logits, dim=-1)
        top_probs, top_indices = torch.topk(probs, beam_width)
        return top_indices.tolist()
    
    def sample_contrastive_search(self, logits: torch.Tensor, penalty_alpha: float = 0.6, 
                                  top_k: int = 4) -> int:
        """对比搜索采样 - 简化实现"""
        top_logits, top_indices = torch.topk(logits, top_k)
        next_token_idx = torch.argmax(top_logits)
        return top_indices[next_token_idx].item()


class SamplingOptimizer:
    """采样优化器 - 提供优化策略和配置建议"""
    
    @staticmethod
    def get_speed_optimized_config() -> dict:
        """获取速度优化配置"""
        return {
            "greedy_threshold": 0.2,      # 更激进的贪婪采样
            "fast_topk_threshold": 2,     # 更小的Top-K范围
            "temperature_threshold": 0.25, # 更低的温度阈值
            "vocab_limit": 30,            # 更小的词汇表限制
        }
    
    @staticmethod
    def get_quality_optimized_config() -> dict:
        """获取质量优化配置"""
        return {
            "greedy_threshold": 0.05,     # 较少使用贪婪采样
            "fast_topk_threshold": 5,     # 较大的Top-K范围
            "temperature_threshold": 0.4,  # 较高的温度阈值
            "vocab_limit": 100,           # 较大的词汇表限制
        }
    
    @staticmethod
    def get_balanced_config() -> dict:
        """获取平衡配置 - 默认推荐"""
        return {
            "greedy_threshold": 0.15,     # 平衡的贪婪采样
            "fast_topk_threshold": 3,     # 平衡的Top-K范围
            "temperature_threshold": 0.3,  # 平衡的温度阈值
            "vocab_limit": 50,            # 平衡的词汇表限制
        }


def benchmark_final_sampler():
    """基准测试最终采样器性能"""
    print("🔬 最终采样器性能基准测试")
    print("=" * 50)
    
    # 创建测试数据
    vocab_size = 151936
    torch.manual_seed(42)
    test_logits = torch.randn(vocab_size) * 2.0
    
    # 创建采样器
    sampler = Sampler(vocab_size)
    
    # 测试不同配置
    configs = [
        ("速度优化", SamplingOptimizer.get_speed_optimized_config()),
        ("平衡配置", SamplingOptimizer.get_balanced_config()),
        ("质量优化", SamplingOptimizer.get_quality_optimized_config()),
    ]
    
    for config_name, config in configs:
        sampler.reset_stats()
        sampler.update_optimization_config(**config)
        
        # 运行测试
        start_time = time.time()
        for _ in range(1000):
            sampler.sample(test_logits, temperature=0.7, top_k=20)
        
        total_time = time.time() - start_time
        stats = sampler.get_performance_stats()
        
        print(f"\n{config_name}:")
        print(f"  总时间: {total_time*1000:.2f}ms")
        print(f"  采样速度: {1000/total_time:.0f} samples/s")
        print(f"  贪婪采样: {stats['greedy_percentage']:.1f}%")
        print(f"  Top-K采样: {stats['topk_percentage']:.1f}%")
        print(f"  温度采样: {stats['temperature_percentage']:.1f}%")


if __name__ == "__main__":
    benchmark_final_sampler()